### ISSUE TITLE
Packet definition design

### ISSUE THREAD

**Issue by Eastrall** [Eastrall/Rhisis#466]
Currently, we have one or multiple interfaces that contains methods definitions for our outgoing packets (Server -> Client). These interfaces are implemented in concrete classes and injected in our actual service systems in order to send packets back to the client.
Example:

```csharp
// IChatPacket.cs
public interface IChatPacket
{
    void SendChatMessage(IPlayerEntity player, string message);
}

// ChatPacket.cs
[Injectable]
public class ChatPacket : IChatPacket
{
    public void SendChatMessage(IPlayerEntity player, string message)
    {
        Write(....);
        Write(....);
        SendToVisible(...);
    }
}
```
Then we inject the `IChatPacket` interface and call the `SendChatMessage()` method with the good parameters and the packet is sent.
This solution is really basic and works pretty fine. But we could improve it and make it more open to future changes.

FLYFF has the possibility to merge packets together and send them in one socket `Send()` operation.

> *If you noticed, all the packets start with a `StartNewMergedPacket(...)` method where the object id and the snapshot type is given. This created a new merged packet, but in fact, only contains one packet.

An idea that could be interesting to implement would be to create concreate classes that will structure the packets and introdue a "Merge()" method that can merge multiple packets.

```csharp
public class ChatPacket : FFPacket
{
    public ChatPacket(IPlayerEntity player, string message)
        : base(player, SnapshotType.CHAT)
    {
        Write(message);
    }
}
```
The code above will create a simple `CHAT` packet with its structure. In order to send it, we could just call:
```csharp
// Assuming we are in a handler or game system with a "player" variable of type "IPlayerEntity"
var chatPacket = new ChatPacket(player, "Hello world!");

player.Connection.SendToVisible(chatPacket);
```

And if we want to merge multiple packet we could introduce the `Merge()` method in the `FFPacket` class:
```csharp
// Assuming we are in a handler or game system with a "player" variable of type "IPlayerEntity"
var chatPacket = new ChatPacket(player, "Hello world!");
var mergedPacketsAsOne = chatPacket.Merge(new OtherPacket(...));

player.Connection.SendToVisible(mergedPacketsAsOne);
```

By doing so, we reduce the amount of dependencies across game systems and allows the possibility to merge packets without having to duplicate code and introduce new methods in the packet factories.

**Comment by Eastrall** [comment]
When sending packets from the server to the client, FLYFF uses a `SNAPSHOT` packet, that allows multiple packets in on single send/receive process. Instead of having a `FFPacket` that merges several packets, let's keep things simple and create a `SnapshotPacket` class that will have the `Merge` method to merge packets in a single one.

```csharp
public class SnapshotPacket : FFPacket
{
    public class SnapshotPacket()
    {
        Write(PacketType.Snapshot);
        Write(0); // Always 0
        Write(0); // Amount of merged packets
    }

    public SnapshotPacket Merge(FFPacket packet)
    {
        // Process merging operation
       return this; // returning the current instance to chain merges
    }
}
```
Usage:
```csharp
var mergedPacket = new SnapshotPacket();

mergedPacket.Merge(new ChatPacket(...))
    .Merge(new BattlePacket(...))
    .Merge(new OtherPacket(...));

player.Connection.Send(mergedPacket);
```

**Comment by Eastrall** [comment]
Implemented in #472
